host/docs/uhd_config_info.1
.TH "uhd_find_devices" 1 "3.9.1" UHD "User Commands"
.SH NAME
uhd_config_info \- USRP Hardware Driver Build Configuration Info
.SH DESCRIPTION
Print build information corresponding to this installation of the USRP
Hardware Driver (UHD).
.LP
The UHD package is the universal hardware driver for Ettus Research
products. The goal is to provide a host driver and API for
current and future Ettus Research products. Users will be able to use
the UHD driver standalone or with 3rd party applications.

.SH SYNOPSIS
.B  uhd_config_info [OPTIONS]

.SH OPTIONS
.IP "Print date this build was compiled:"
--boost-version
.IP "Print Boost version used:"
--build-date
.IP "Print C compiler used:"
--c-compiler
.IP "Print C++ compiler used:"
--cxx-compiler
.IP "Print C compile flags:"
--c-flags
.IP "Print C++ compile flags:"
--cxx-flags
.IP "Print UHD components included in this build (comma-delimited):"
--enabled-components
.IP "Print default install prefix for this build:"
--install-prefix
.IP "Print libusb version used"
--libusb-version
.IP "Print all information listed above:"
--print-all
.IP "Print UHD version:"
--version
.IP "This help information:"
--help

.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html

.SH AUTHOR
This manual page was written by Nicholas Corgan
for the Debian project (but may be used by others).

.SH COPYRIGHT
Copyright (c) 2015 National Instruments Corp.
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.009.002
* E310: Added support for Speedgrade 3
* B205mini: Added support
* E310: Fixed reference counting bug
* B210: Fixed external clock reference bug for devices using ADF4002 PLLs
* B210: Fixed codec loopback test
* OctoClock: Fixed firmware burning on Windows
* B2XX, E3XX, X3XX: Easier time-syncing features. Fixes bug where B210s would
  only run after issuing set_time_unknown_pps().
* X3XX: Fixed bug for IQ imbalance correction
* E310: DRAM testbenching
* Docs/Manual: Many updates and fixes

## 003.009.001
* B200mini: Updated udev rules, removed DCM
* B200: Better USB error messages
* Cores: Fixed CORDIC scaling issue on all devices that use
  dsp_core_3000
* X300: Fixed GPIO issue
* Examples: Added PPS source option
* Docs: Multiple manual updates

## 003.009.000
* X300: Updated DAC ctrl, FPGA toolchain is now entirely Vivado,
  improved master clock controls, added ADC self-cal capability,
  prepared for revisions 7 and 8, fixed flow control issue which
  could cause device to hang when receiving too many overruns
* B2XX: Auto clock rate setting, added PID/VID pairs to support
  all B2XX- and derivatives, added temperature sensor, improved
  DC offset and IQ imbalance correction, added AGC support,
  support for FPGPIO connector on Rev6+ boards, full clock range support,
  updated FX3 firmware (side-channel logging capabilities, updated
  tx voltage swing, better configurability), default tick rate now
  16 MHz, added B200mini support
* E3XX: Added temperature sensor, FPGA toolchain is now entirely Vivado,
  improved DC offset and IQ imbalance correction, added AGC support,
  improved FPGA capture interface robustness for RFIC, make frame
  sizes configurable, replaced GPS control code with gpsd interfacing
  capabilities
* Octoclock: Fixed bootloader + ethernet capabilities
* Compilers: Supported MSVC versions are now 2012, 2013, 2015
  (dropped 2010 support), added MinGW capabilities
* Documentation: Many minor fixes and updates, merged all the
  info from code.ettus.com
* UHD: Added sid_t, CHDR-specific transports now get their own
  (un)packer codes, fixed a lot of compiler warnings, added
  filter API (currently available for AD9361 frontend), added
  soft-register API, replaced Cheetah with Mako, full Py3k
  compliance, updated images downloader tool (now is one tool
  for all devices), CMake minimum version is now 2.8, refactored
  general AD9361 peripheral management, refactored most core
  control management, added usb_error type (used by B2xx devices),
  better exception handling at runtime, added C wrapper API,
  new dependency: python-requests
* C API: Added to UHD (wraps C++ calls in C)
* multi_usrp: Added normalized gain setters/getters, IQ imbalance
  + DC offset correction API, filter API
* Converters: Converter symbols now exported, better logging,
  removed ORC dependency, added u8 converters
* Examples: Whitespace- and other cleanup, multi-channel fixes for
  some examples
* Utils: Read more property tree types from the command line
* Tools: kitchen sink updated, added mega_fft

## 003.008.005
* E3xx: Added support for battery-based E3xx device (E312),
  get_freq() for network mode, fixed GPS time initialization bug
* AD9361-based devices: Fixed frequency readback bug
* B200: Fixed DCM issues, better loopback failure handling,
  fix erroneous warning for custom clock rates
* X3x0: Better warnings for clock reate / ref freq
* multi_usrp: Added define for GPIO capabilities (enables
  exposure in GNU Radio)
* UHD: sc16->sc16 SSE converter
* Manual: Multiple minor updates, FPGA manual improvements,
* Build System: Fixed builds on some Windows platforms, removed
  stray prints

## 003.008.004
* B200: Fixed EEPROM writing bug, updated images for B200 Rev5/6
* E300: GPS antenna power defaults to staying on, GPS time used as
  default if available
* UBX: PDF frequency fix on X300
* USRP2: Bugfix that allows latest UHD to work with USRP2
* Documentation: Many fixes for E300 section, added FPGA manual

## 003.008.003
* UBX: Fixed phase synchronization issues
  (Related changes: Change X300 daughterboard frequency, increase
   N210 FIFO depth)
* Fixed many compiler warnings
* B200: Fixed timing issues, fixed tick rate issue, stabilized
  operations at high clock rates
* X300: Improved phase alignment across devices
* CMake: Build fixes
* E300: Flow control fix


## 003.008.002
* CMake: Introducing named versions for dev branches, enable static libs,
* E300: Docs updates, compat number bump, VCTCXO updates,
  more status bits for ref locking, fixed serial number length fix,
  RSSI Sensor
* B200: RSSI sensor
* AD9361: Better handling of different interpolation rates in FIR,
  fix for gain value truncation, removed gain value offset
* UBX: Added drivers
* Manual/Docs: Numerous updates, minimum compiler versions now specified,
* Converters: Multiple fixes for sc12
* Examples: Fixes to txrx_loopback_to_file
* Path handling vastly improved, made more consistent
* Minor UHD fixes

## 003.008.001
* B2x0: Fixed PLL settings, Fixed external ref selection, serialized
  streamer setup (thread-safety)
* X3x0: Fixed flow control issue, improved DAC ctrl + init logic,
  Fixed I/Q alignment issue
* Generation-3 devices: Fixed LED registers
* UHD: Improved tuning logic for manual tunes
* Tools: Multiple kitchen sink fixes, coloured output
* Examples: Multiple bugfixes (multi-channel ops)
* Docs/Manual: Multiple fixes, E310 panel images

## 003.008.000
* Added E310 support
* B200/B210: Moved AD9361 controls from firmware to host
* Added several tools: ZPU dissector, improved CHDR dissector,
  kitchen sink, B200/B210 USB debugging utility, latency
  measurement tool.
* Reorganized firmware/ directory structure. Refactored some
  firmware.
* Removed FPGA sources, is now in own repository (submoduled).
* Cleaned up command line arguments for some tools
* Added math namespace, plus a unified float comparison infrastructure

host/utils/uhd_config_info.cpp
//
// Copyright 2015 National Instruments Corp.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include <uhd/build_info.hpp>
#include <uhd/version.hpp>
#include <uhd/utils/safe_main.hpp>

#include <boost/format.hpp>
#include <boost/program_options.hpp>

namespace po = boost::program_options;

int UHD_SAFE_MAIN(int argc, char* argv[]) {
    // Program Options
    po::options_description desc("Allowed Options");
    desc.add_options()
        ("build-date",         "Print build date")
        ("c-compiler",         "Print C compiler")
        ("cxx-compiler",       "Print C++ compiler")
        ("c-flags",            "Print C compiler flags")
        ("cxx-flags",          "Print C++ compiler flags")
        ("enabled-components", "Print built-time enabled components")
        ("install-prefix",     "Print install prefix")
        ("libusb-version",     "Print libusb version")
        ("print-all",          "Print everything")
        ("version",            "Print this UHD build's version")
        ("help",               "Print help message")
    ;

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // Print the help message
    if(vm.count("help") > 0) {
        std::cout << boost::format("UHD Config Info - %s") % desc << std::endl;
        return EXIT_FAILURE;
    }

    bool print_all = (vm.count("print-all") > 0);

    if(vm.count("version") > 0 or print_all) {
        std::cout << "UHD " << uhd::get_version_string() << std::endl;
    }
    if(vm.count("build-date") > 0 or print_all) {
        std::cout << "Build date: " << uhd::build_info::build_date() << std::endl;
    }
    if(vm.count("c-compiler") > 0 or print_all) {
        std::cout << "C compiler: " << uhd::build_info::c_compiler() << std::endl;
    }
    if(vm.count("cxx-compiler") > 0 or print_all) {
        std::cout << "C++ compiler: " << uhd::build_info::cxx_compiler() << std::endl;
    }
    if(vm.count("c-flags") > 0 or print_all) {
        std::cout << "C flags: " << uhd::build_info::c_flags() << std::endl;
    }
    if(vm.count("cxx-flags") > 0 or print_all) {
        std::cout << "C++ flags: " << uhd::build_info::cxx_flags() << std::endl;
    }
    if(vm.count("enabled-components") > 0 or print_all) {
        std::cout << "Enabled components: " << uhd::build_info::enabled_components() << std::endl;
    }
    if(vm.count("install-prefix") > 0 or print_all) {
        std::cout << "Install prefix: " << uhd::build_info::install_prefix() << std::endl;
    }
    if(vm.count("boost-version") > 0 or print_all) {
        std::cout << "Boost version: " << uhd::build_info::boost_version() << std::endl;
    }
    if(vm.count("libusb-version") > 0 or print_all) {
        std::string _libusb_version = uhd::build_info::libusb_version();
        std::cout << "Libusb version: " << (_libusb_version.empty() ? "N/A" : _libusb_version) << std::endl;
    }

    return EXIT_SUCCESS;
}
